evmdis/main.go

```diff
@@ -2,32 +2,98 @@ package main
+
+var swarmHashProgramTrailer = [...]byte{0x00, 0x29}
+var swarmHashHeader = [...]byte{0xa1, 0x65}
+
 func main() {
+
+	withSwarmHash := flag.Bool("swarm", true, "solc adds a reference to the Swarm API description to the generated bytecode, if this flag is set it removes this reference before analysis")
+	ctorMode := flag.Bool("ctor", false, "Indicates that the provided bytecode has construction(ctor) code included. (needs to be analyzed seperatly)")
+	logging := flag.Bool("log", false, "print logging output")
+
+	flag.Parse()
+
+	if !*logging {
+		log.SetOutput(ioutil.Discard)
+	}
+
 	hexdata, err := ioutil.ReadAll(os.Stdin)
 	if err != nil {
 		panic(fmt.Sprintf("Could not read from stdin: %v", err))
 	}
 
-	bytecode := make([]byte, hex.DecodedLen(len(hexdata)))
+	bytecodeLength := uint64(hex.DecodedLen(len(hexdata)))
+	bytecode := make([]byte, bytecodeLength)
+
 	hex.Decode(bytecode, hexdata)
 
-	program := evmdis.NewProgram(bytecode)
-	if err := evmdis.PerformReachingAnalysis(program); err != nil {
-		panic(fmt.Sprintf("Error performing reaching analysis: %v", err))
+	// detect swarm hash and remove it from bytecode, see http://solidity.readthedocs.io/en/latest/miscellaneous.html?highlight=swarm#encoding-of-the-metadata-hash-in-the-bytecode
+	if bytecode[bytecodeLength-1] == swarmHashProgramTrailer[1] &&
+		bytecode[bytecodeLength-2] == swarmHashProgramTrailer[0] &&
+		bytecode[bytecodeLength-43] == swarmHashHeader[0] &&
+		bytecode[bytecodeLength-42] == swarmHashHeader[1] &&
+		*withSwarmHash {
+		bytecodeLength -= swarmHashLength // remove swarm part
 	}
-	evmdis.PerformReachesAnalysis(program)
-	evmdis.CreateLabels(program)
-	if err := evmdis.BuildExpressions(program); err != nil {
-		panic(fmt.Sprintf("Error building expressions: %v", err))
+
+	program := evmdis.NewProgram(bytecode[:bytecodeLength])
+	AnalyzeProgram(program)
+
+	if *ctorMode {
+		var codeEntryPoint = FindNextCodeEntryPoint(program)
+
+		if codeEntryPoint == 0 {
+			panic("no code entrypoint found in ctor")
+		} else if codeEntryPoint >= bytecodeLength {
+			panic("code entrypoint outside of currently available code")
+		}
+
+		ctor := evmdis.NewProgram(bytecode[:codeEntryPoint])
+		code := evmdis.NewProgram(bytecode[codeEntryPoint:bytecodeLength])
+
+		AnalyzeProgram(ctor)
+		fmt.Println("# Constructor part -------------------------")
+		PrintAnalysisResult(ctor)
+
+		AnalyzeProgram(code)
+		fmt.Println("# Code part -------------------------")
+		PrintAnalysisResult(code)
+
+	} else {
+		PrintAnalysisResult(program)
+	}
+}
+
+func FindNextCodeEntryPoint(program *evmdis.Program) uint64 {
+	var lastPos uint64 = 0
+	for _, block := range program.Blocks {
+		for _, instruction := range block.Instructions {
+			if instruction.Op == evmdis.CODECOPY {
+				var expression evmdis.Expression
+
+				instruction.Annotations.Get(&expression)
+
+				if i, err := strconv.ParseUint(expression.(*evmdis.InstructionExpression).Arguments[1].String()[2:], 16, 32); err == nil {
```

Comment:
I'd still rather that you added an `Eval()` method to Expression, that returns a bigint when it knows the value statically, and nil when it doesn't, than rely on casting things to string and back. String representation is strictly for showing to users, not for internal handling.